services/121-service/src/payments/fsp-integration/safaricom/safaricom.service.ts

```diff
@@ -0,0 +1,204 @@
+import { SafaricomRequestEntity } from './safaricom-request.entity';
+import { SafaricomApiService } from './safaricom.api.service';
+
+@Injectable()
+export class SafaricomService {
+  @InjectRepository(SafaricomRequestEntity)
+  private readonly safaricomRequestRepository: Repository<SafaricomRequestEntity>;
+  @InjectRepository(TransactionEntity)
+  private readonly transactionRepository: Repository<TransactionEntity>;
+
+  public constructor(
+    private readonly safaricomApiService: SafaricomApiService,
+    private readonly transactionsService: TransactionsService,
+  ) {}
+
+  public async sendPayment(
+    paymentList: PaPaymentDataDto[],
+    programId: number,
+    paymentNr: number,
+  ): Promise<FspTransactionResultDto> {
+    const fspTransactionResult = new FspTransactionResultDto();
+    fspTransactionResult.paList = [];
+    fspTransactionResult.fspName = FspName.safaricom;
+
+    await this.safaricomApiService.authenticate();
+
+    for (const payment of paymentList) {
+      const payload = this.createPayloadPerPa(payment, paymentNr);
+
+      const paymentRequestResultPerPa = await this.sendPaymentPerPa(
+        payload,
+        payment.referenceId,
+      );
+      fspTransactionResult.paList.push(paymentRequestResultPerPa);
+      // Storing the per payment so you can continiously seed updates of transactions in HO-Portal
+      await this.transactionsService.storeTransactionUpdateStatus(
+        paymentRequestResultPerPa,
+        programId,
+        paymentNr,
+      );
+
+      await this.processSafaricomRequest(payload, paymentRequestResultPerPa);
+    }
+    return fspTransactionResult;
+  }
+
+  public createPayloadPerPa(payment, paymentNr): SafaricomTransferPayload {
+    function randomValueHex(len: number): string {
+      return crypto
+        .randomBytes(Math.ceil(len / 2))
+        .toString('hex') // convert to hexadecimal format
+        .slice(0, len)
+        .toUpperCase(); // return required number of characters
+    }
+
+    const payload = {
+      InitiatorName: process.env.SAFARICOM_INITIATORNAME,
+      SecurityCredential: process.env.SAFARICOM_SECURITY_CREDENTIAL,
+      CommandID: 'BusinessPayment',
+      Amount: payment.transactionAmount,
+      PartyA: process.env.SAFARICOM_PARTY_A,
+      PartyB: payment.paymentAddress,
+      Remarks: `Payment ${paymentNr}`,
+      QueueTimeOutURL: process.env.SAFARICOM_QUEUETIMEOUT_URL,
+      ResultURL: process.env.SAFARICOM_RESULT_URL,
+      Occassion: payment.referenceId,
+      OriginatorConversationID: `232323_KCBOrg_${randomValueHex(20)}`,
+      IDType: process.env.SAFARICOM_IDTYPE,
+      IDNumber: '123789',
+    };
+
+    return payload;
+  }
+
+  public async sendPaymentPerPa(
+    payload: any,
+    referenceId: string,
+  ): Promise<PaTransactionResultDto> {
+    // A timeout of 100ms to not overload safaricom server
+    await new Promise((r) => setTimeout(r, 2000));
```

Comment:
This timeout is still on 2 seconds. Lets take a more reasonable amount or remove it.

Reply:
Removed.